### PR TITLE
move test waitForBucketReadAccess before browser checkBucketReadAccess

### LINKED ISSUE #146: GAWB #3711: UI incorrectly states google bucket read access was disabled
https://broadinstitute.atlassian.net/browse/GAWB-3711 

Scope: Depending on results of investigation for reasoning of this bug, may or may not fall on FC App for completion.

### PR DESCRIPTION
**Pull request by davidangb**
attempting to fix DataBiosphere/firecloud-app#146 / GAWB-3711.

In this test, the `waitForBucketReadAccess` pause was *after* the test opened the workspace in the browser. The UI triggers its own `checkBucketReadAccess` call when the browser opens the workspace. Thus, it was possible for this test to get a 403 on `checkBucketReadAccess` in the UI, then continue to wait in the test code until the test's own `waitForBucketReadAccess` succeeded, then continue to drive the browser - which would result in an error.

This is a relatively infrequent test error, though it does happen repeatedly. In the last ~100 dev-deploy test runs, I saw this error once.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/DataSpec.scala [hunk 1]
```diff
@@ -35,10 +35,10 @@ class DataSpec extends FreeSpec with ParallelTestExecution with WebBrowserSpec w
 
           withWebDriver { implicit driver =>
             withSignIn(owner) { _ =>
+              api.workspaces.waitForBucketReadAccess(billingProject, workspaceName)
```
Comment (aweng98) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/DataSpec.scala]: If I was doing this, I would move `waitForBucketReadAccess` to line # 35, before `withWebDriver`. I'm skeptical this change is going to make any difference. But it worth a shot.

**Comment by aweng98** [comment]
Might have other places where `waitForBucketReadAccess` is used inside of `withWebDriver` fixture.

**Review by aweng98 — APPROVED** [review]

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/DataSpec.scala (hunk 1 above)
Comment (davidangb) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/DataSpec.scala]: the pattern of `waitForBucketReadAccess` inside of `withWebDriver` is present in about 6 other tests, so it felt safe - happy to move it, though. The majority of tests beyond those 6 have it outside `withWebDriver`.

**Comment by davidangb** [comment]
and @aweng98 for more detailed explanation: the problem I am solving here is a misunderstanding of how/when the UI checks if the user has bucket access. The original code looked like this, with my comments inline:
```scala
withCleanBillingProject(owner) { billingProject =>
        /**************************************************
          create a new workspace for this test. ACLs for this workspace's bucket do not take effect immediately;
            it takes some time for Google to reach consistency.
          **************************************************/
        withWorkspace(billingProject, "DataSpec_launch_workflow", aclEntries = List(AclEntry(reader.email, WorkspaceAccessLevel.Reader))) { workspaceName =>
          api.importMetaData(billingProject, workspaceName, "entities", testData.participantEntity)
          api.methodConfigurations.copyMethodConfigFromMethodRepo(billingProject, workspaceName, SimpleMethodConfig.configNamespace,
            SimpleMethodConfig.configName, SimpleMethodConfig.snapshotId, SimpleMethodConfig.configNamespace, methodConfigName)

          withWebDriver { implicit driver =>
            withSignIn(owner) { _ =>
              /**************************************************
                open the workspace in the browser.
                  this loads the workspace details and triggers an ajax call to checkBucketReadAccess to see if the user
                  has permissions to the workspace bucket. The results of both of these ajax calls are cached im the UI
                  and reused by each of the workspace's tabs.
                At this moment, the ACLs from {withWorkspace} may not have reached consistency, and the {owner}'s call
                  to checkBucketReadAccess may return a 403. This 403 will be remembered.
                **************************************************/
              val workspaceDataTab = new WorkspaceDataPage(billingProject, workspaceName).open
              val headers1 = List("participant_id")
              eventually { workspaceDataTab.dataTable.readColumnHeaders shouldEqual headers1 }
              /**************************************************
                Run our test utility to pause the rest of the test until {owner}'s call to checkBucketReadAccess succeeds.
                  This is too late, since the {WorkspaceDataPage.open} command has already executed.
                **************************************************/
              api.workspaces.waitForBucketReadAccess(billingProject, workspaceName)
              val methodConfigDetailsPage = new WorkspaceMethodConfigDetailsPage(billingProject, workspaceName, SimpleMethodConfig.configNamespace, methodConfigName).open
              /**************************************************
                Attempt to launch an analysis. In the UI, the "Launch Analysis" button is enabled or disabled based on the
                  cached result of {owner}'s call to checkBucketReadAccess that was triggered by {WorkspaceDataPage.open}.
                Even though the {launchAnalysis} command happens after {waitForBucketReadAccess}, it is using data from
                  before {waitForBucketReadAccess}, and therefore this test can fail.
                **************************************************/
              val submissionTab = methodConfigDetailsPage.launchAnalysis(SimpleMethodConfig.rootEntityType, testData.participantId, "", true)
              submissionTab.waitUntilSubmissionCompletes()
              assert(submissionTab.verifyWorkflowSucceeded())
              workspaceDataTab.open
```

**Review by jmthibault79 — APPROVED** [review]

**Pull request merged**